internalshared/configutil/normalize.go

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package configutil
+
+import (
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+)
+
+// NormalizeAddr takes an address as a string and returns a normalized copy.
+// If the addr is a URL, IP Address, or host:port address that includes an IPv6
+// address, the normalized copy will be conformant with RFC-5942 §4
+// See: https://rfc-editor.org/rfc/rfc5952.html
+func NormalizeAddr(u string) string {
+	if u == "" {
+		return ""
+	}
+
+	var ip net.IP
+	var port string
+	bracketedIPv6 := false
+
+	// Try parsing it as a URL
+	pu, err := url.Parse(u)
+	if err == nil {
+		// We've been given something that appears to be a URL. See if the hostname
+		// is an IP address
+		ip = net.ParseIP(pu.Hostname())
+	} else {
+		// We haven't been given a URL. Try and parse it as an IP address
+		ip = net.ParseIP(u)
+		if ip == nil {
+			// We haven't been given a URL or IP address, try parsing an IP:Port
+			// combination.
+			idx := strings.LastIndex(u, ":")
+			if idx > 0 {
+				// We've perhaps received an IP:Port address
+				addr := u[:idx]
+				port = u[idx+1:]
+				if strings.HasPrefix(addr, "[") && strings.HasSuffix(addr, "]") {
+					addr = strings.TrimPrefix(strings.TrimSuffix(addr, "]"), "[")
+					bracketedIPv6 = true
+				}
+				ip = net.ParseIP(addr)
```

Comment:
For the following IP address: `2001:db8::68`, wouldn't this code currently assign to `addr` and parse `2001:db8:` as an IP?

Reply:
I'm not sure I understand. `2001:db8::68` is a valid IPv6 address so it would have been parsed successfully on line 34 and the branch scope that contains this code would be skipped entirely.

Reply:
Right! Yeah, I'm dumb, sorry. Resolving!